src/wallet/bdb.cpp

```diff
@@ -103,6 +122,8 @@ void BerkeleyEnvironment::Close()
         if (database.m_db) {
             database.m_db->close(0);
             database.m_db.reset();
+            LOCK(cs_db);
+            g_fileids.erase(database.m_file_path);
```

Comment:
Ah, I see how that could theoretically happen. I did a bit of testing and wasn't able to trigger a spurious duplicate fileid error. I believe that's because closing database A also closed database B since nothing was using it. So to get this error, B needs to be in use.

The reason this was failing tests was because I corrected a mistake in the patch. You have `g_fileids.erase(m_file_path);` which should actually be `erased = g_fileids.erase(m_file_path);`. Otherwise the following `assert(erased == 1);` doesn't do anything. But because `BerkeleyEnvironment::Close` removes the fileid before the destructor is called, this assertion will fail.

So instead, I've dropped that assertion but still kept the `g_fileids.erase(m_file_path);`.